waspc/examples/todoApp/src/server/apis.ts

```diff
@@ -0,0 +1,7 @@
+import { Request, Response } from '@wasp/types'
+import { fooBarContext } from '@wasp/apis/types'
+
+export function fooBar(req: Request, res: Response, context: fooBarContext) {
```

Comment:
What do people think of providing a context type (`waspc/data/Generator/templates/server/src/apis/types.ts`) instead of some wrapper function?

Reply:
Also, should we uppercase the start of the typename?

Reply:
@sodic will probably be the best person to give advice here!

Reply:
Yep, the convention is to uppercase the types in Typescript. 

> What do people think of providing a context type (waspc/data/Generator/templates/server/src/apis/types.ts) instead of some wrapper function?

Could you provide code examples, I'm not 100% I understand what you mean?

Reply:
Wrapper function was incorrect. I meant I am not creating a function-level type that provides the types the context, amongst other things, like Queries: https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/server/src/queries/types.ts Here, it's just for the params.

I can uppercase the first letter too 👍🏻

Reply:
Made type uppercase.

Reply:
Yes, it's better to expose the type of the entire function in this case. Just to be clear, this is what I understood you meant:
```typescript
import { FooBar } from '@wasp/apis/types'

export const fooBar: FooBar = (req, res, context) => {
  // ...
}
```
We always need a function of type `(req: Request, res: Response, context: fooBarContext) => void`:
  - If they specify a function of type `(req: string, res: Response, context: fooBarContext) => void`, they'll run into errors.
  - If they specify a function of type `(req: Request, res: SomethingElse, context: fooBarContext) => void`, they'll run into errors.
  - etc.
 
On the other hand, if we just tell them to use the type `FooBar` that's equal to `(req: string, res: Response, context: fooBarContext) => unknown`, they can't make such errors.
Also, it's only a single import that assigns correct types to everything they need - fewer imports, fewer things to remember, less code, smaller chance of making a mistake.
Finally, it's more similar to the API we already have.

The only problem is, `Response` and `Request` are generic types (and they can take quite a few type arguments):
```typescript
interface Request<
    P = core.ParamsDictionary,
    ResBody = any,
    ReqBody = any,
    ReqQuery = core.Query,
    Locals extends Record<string, any> = Record<string, any>
> extends core.Request<P, ResBody, ReqBody, ReqQuery, Locals> {}

interface Response<
    ResBody = any,
    Locals extends Record<string, any> = Record<string, any>
> extends core.Response<ResBody, Locals> {}
```
By hiding them behind `FooBar`, we take away the option to specify the argument types from our users.

The good new is, if you take a closer look, you'll see several of these type arguments are the same (e..g, `ResBody`). In other words, these two types (`Request` and `Response`) are coupled.
To prevent mistakes such a coupling could cause (i.e., to make illegal states unrepresentable), Express also exposes a function type for the handler:
```typescript
interface RequestHandler<
    P = core.ParamsDictionary,
    ResBody = any,
    ReqBody = any,
    ReqQuery = core.Query,
    Locals extends Record<string, any> = Record<string, any>
> extends core.RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals> {}
```
Which is pretty  much exactly what we should be using - our `FooBar` type should decorate `RequestHandler`.

Reply:
Updated to use a function type, thanks for the help @sodic!